File: backend/src/main/java/com/agmsentinel/service/FeatureService.java
package com.agmsentinel.service;

import com.agmsentinel.model.FeatureFlag;
import com.agmsentinel.repository.FeatureFlagRepository;
import com.agmsentinel.security.Feature;
import com.agmsentinel.security.JwtService;
import io.jsonwebtoken.Claims;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.event.EventListener;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

/**
 * Whether a feature is on, and who may use it.
 *
 * <p>Reads are the hot path — every guarded request asks — so the resolved state is held in memory
 * and refreshed on write. A database round trip per request to answer "is this switched on" would
 * make the flag system itself the slowest thing about the features it guards.
 *
 * <p>The store holds only what an admin has changed. Anything without a row behaves as the
 * catalogue default, which is what lets a new feature ship dark without anyone having to insert a
 * row first.
 */
@Service
public class FeatureService {

    private static final Logger log = LoggerFactory.getLogger(FeatureService.class);

    private final FeatureFlagRepository flags;

    /** Resolved state, rebuilt on every write. Volatile map reference, replaced wholesale. */
    private volatile Map<Feature, Resolved> state = Map.of();

    /** One feature's effective configuration. */
    public record Resolved(Feature feature, boolean enabled, Set<String> allowedRoles,
                           String updatedBy, boolean customised) { }

    public FeatureService(FeatureFlagRepository flags) {
        this.flags = flags;
    }

    /**
     * Load the stored overrides once the context is up.
     *
     * <p>Guarded: a feature-flag table that cannot be read must not stop the application starting.
     * Falling back to the catalogue defaults means the app comes up behaving exactly as it would
     * with no overrides — degraded, but running and diagnosable.
     */
    @EventListener(ApplicationReadyEvent.class)
    public void loadOnStartup() {
        try {
            refresh();
            long on = state.values().stream().filter(Resolved::enabled).count();
            log.info("Feature flags loaded: {} of {} enabled.", on, Feature.values().length);
        } catch (RuntimeException ex) {
            log.error("Could not read feature flags; falling back to defaults. {}", ex.getMessage());
            this.state = defaults();
        }
    }

    @Transactional(readOnly = true)
    public void refresh() {
        Map<String, FeatureFlag> stored = new LinkedHashMap<>();
        for (FeatureFlag flag : flags.findAll()) {
            stored.put(flag.getKey().toUpperCase(), flag);
        }
        Map<Feature, Resolved> resolved = new LinkedHashMap<>();
        for (Feature feature : Feature.values()) {
            FeatureFlag flag = stored.get(feature.key());
            if (flag == null) {
                resolved.put(feature, new Resolved(feature, feature.enabledByDefault(),
                        feature.defaultRoles(), null, false));
            } else {
                Set<String> roles = flag.getAllowedRoles().isEmpty()
                        ? feature.defaultRoles() : Set.copyOf(flag.getAllowedRoles());
                resolved.put(feature, new Resolved(feature, flag.isEnabled(), roles,
                        flag.getUpdatedBy(), true));
            }
        }
        this.state = resolved;
    }

    private Map<Feature, Resolved> defaults() {
        Map<Feature, Resolved> map = new LinkedHashMap<>();
        for (Feature feature : Feature.values()) {
            map.put(feature, new Resolved(feature, feature.enabledByDefault(),
                    feature.defaultRoles(), null, false));
        }
        return map;
    }

    private Resolved resolve(Feature feature) {
        Resolved current = state.get(feature);
        if (current != null) return current;
        // Asked before startup finished, or for a feature added since the last refresh.
        return new Resolved(feature, feature.enabledByDefault(), feature.defaultRoles(), null, false);
    }

    // ---- asking --------------------------------------------------------------

    public boolean isEnabled(Feature feature) {
        return resolve(feature).enabled();
    }

    /** Enabled, and permitted for at least one role this caller holds. */
    public boolean isAvailableTo(Feature feature, Set<String> roles) {
        Resolved resolved = resolve(feature);
        if (!resolved.enabled()) return false;
        // ADMIN is a superset throughout the application; a switched-on feature is never hidden
        // from the person who administers the switches.
        if (roles.contains("ADMIN")) return true;
        return resolved.allowedRoles().stream().anyMatch(roles::contains);
    }

    /** Every feature, with its effective state — what the admin screen renders. */
    public List<Resolved> all() {
        return Feature.values().length == state.size()
                ? List.copyOf(state.values())
                : List.copyOf(defaults().values());
    }

    /** The subset available to the current caller — what the SPA uses to decide what to show. */
    public List<Feature> availableToCurrentUser() {
        Set<String> roles = currentRoles();
        return java.util.Arrays.stream(Feature.values())
                .filter(f -> isAvailableTo(f, roles))
                .toList();
    }

    /**
     * Refuse the request when the feature is off or not permitted for this caller.
     *
     * <p>404 rather than 403 when it is switched off. A disabled feature should not be discoverable
     * by probing — "you are not allowed" confirms it exists, which is a different statement from
     * "there is nothing here".
     */
    public void require(Feature feature) {
        Resolved resolved = resolve(feature);
        if (!resolved.enabled()) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND,
                    "That feature is not enabled on this deployment.");
        }
        if (!isAvailableTo(feature, currentRoles())) {
            throw new ResponseStatusException(HttpStatus.FORBIDDEN,
                    "Your role does not have access to " + feature.label() + ".");
        }
    }

    // ---- changing ------------------------------------------------------------

    /**
     * Turn a feature on or off, and optionally narrow which roles may use it.
     *
     * <p><b>Null and empty mean different things.</b> Null is "I am not changing the roles" — the
     * caller is only flipping the switch — and keeps whatever is already configured. Empty is an
     * explicit "no role may use this", and is stored as such.
     *
     * <p>They were once treated the same, both falling back to the catalogue defaults. That made
     * unticking the last role silently re-tick every default one, which reads as the screen undoing
     * the click. An admin narrowing a feature to nothing has said something deliberate, and the
     * honest response is to store it — the feature is then effectively off for everyone but ADMIN,
     * which is exactly what they asked for and is visible on the screen.
     */
    @Transactional
    public Resolved update(Feature feature, boolean enabled, Set<String> allowedRoles, String actor) {
        Set<String> roles = allowedRoles == null
                ? resolve(feature).allowedRoles()
                : Set.copyOf(allowedRoles);

        FeatureFlag flag = flags.findById(feature.key()).orElse(null);
        if (flag == null) {
            flag = new FeatureFlag(feature.key(), enabled, roles, actor);
        } else {
            flag.setEnabled(enabled);
            flag.setAllowedRoles(roles);
            flag.setUpdatedBy(actor);
        }
        flags.save(flag);
        refresh();
        log.info("Feature {} {} by {} (roles: {}).", feature.key(),
                 enabled ? "enabled" : "disabled", actor, roles);
        return resolve(feature);
    }

    /**
     * Apply one decision to every feature at once.
     *
     * <p>Exists because configuring sixteen features one switch at a time is the common case for a
     * fresh deployment — "turn everything on for everyone" is a setup step, not a fine adjustment,
     * and doing it by hand invites missing one.
     *
     * <p><b>Granting every role cannot escalate anything.</b> Roles on a feature narrow access; they
     * never widen it, because {@code SecurityConfig} is checked first and independently. A
     * shareholder listed against a moderator-only route still gets a 403 from Spring Security. That
     * is what makes a blunt bulk action safe to offer.
     *
     * <p>Written as one transaction and one refresh, so the cached state is never half-applied.
     *
     * @param roles null to leave each feature's roles as they are — see {@link #update}
     */
    @Transactional
    public List<Resolved> updateAll(boolean enabled, Set<String> roles, String actor) {
        for (Feature feature : Feature.values()) {
            Set<String> effective = roles == null ? resolve(feature).allowedRoles() : Set.copyOf(roles);
            FeatureFlag flag = flags.findById(feature.key()).orElse(null);
            if (flag == null) {
                flag = new FeatureFlag(feature.key(), enabled, effective, actor);
            } else {
                flag.setEnabled(enabled);
                flag.setAllowedRoles(effective);
                flag.setUpdatedBy(actor);
            }
            flags.save(flag);
        }
        refresh();
        log.warn("ALL {} features {} by {}{}.", Feature.values().length,
                 enabled ? "enabled" : "disabled", actor,
                 roles == null ? "" : " for roles " + roles);
        return all();
    }

    /**
     * Drop every override, returning the whole deployment to how it ships.
     *
     * <p>The way back from a bulk change that went too far, and the reason
     * {@link #updateAll} is safe to press.
     */
    @Transactional
    public List<Resolved> resetAll(String actor) {
        flags.deleteAll();
        refresh();
        log.warn("All feature overrides cleared by {} — every feature is back to its default.", actor);
        return all();
    }

    /** Drop the override so the feature returns to its catalogue default. */
    @Transactional
    public Resolved reset(Feature feature, String actor) {
        flags.findById(feature.key()).ifPresent(flags::delete);
        refresh();
        log.info("Feature {} reset to its default by {}.", feature.key(), actor);
        return resolve(feature);
    }

    // ---- caller ---------------------------------------------------------------

    /**
     * Every role the caller holds.
     *
     * <p>Read from the verified token's claims where available, falling back to granted authorities.
     * Both describe the same thing; the claims are simply closer to the source.
     */
    private Set<String> currentRoles() {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        if (auth == null) return Set.of();
        if (auth.getCredentials() instanceof Claims claims) {
            return Set.copyOf(JwtService.rolesOf(claims));
        }
        return auth.getAuthorities().stream()
                .map(a -> a.getAuthority())
                .map(a -> a.startsWith("ROLE_") ? a.substring(5) : a)
                .collect(java.util.stream.Collectors.toUnmodifiableSet());
    }

    public Optional<Feature> lookup(String key) {
        return Feature.of(key);
    }
}

File: frontend/src/app/services/feature.service.ts
import { HttpClient, HttpContext } from '@angular/common/http';
import { Injectable, computed, inject, signal } from '@angular/core';
import { Observable, tap } from 'rxjs';
import { environment } from '../../environments/environment';
import { SILENT } from './loading.service';
import { AuthService } from './auth.service';

/** Keys must match the backend `Feature` enum — it is the catalogue. */
export type FeatureKey =
  | 'VIDEO_LIBRARY'
  | 'VIDEO_ENGAGEMENT'
  | 'VIDEO_DOWNLOAD'
  | 'VIDEO_CHAPTERS'
  | 'VIDEO_WATCH_TRACKING'
  | 'LOUNGE_CHAT'
  | 'AI_DRAFTING'
  | 'MEETINGS'
  | 'SEMANTIC_SEARCH'
  | 'HELP_WIDGET'
  | 'AUTO_TRANSCRIPTION'
  | 'VOTING'
  | 'QUORUM'
  | 'CLUSTER_UPVOTE'
  | 'CLUSTER_CURATION'
  | 'RUN_OF_SHOW'
  | 'MEETING_REPORTS'
  | 'ATTENDEE_BOARD';

/** One feature as the admin screen sees it. */
export interface FeatureView {
  key: FeatureKey;
  label: string;
  description: string;
  enabled: boolean;
  allowedRoles: string[];
  /** False while it is still on its shipped default — nobody has touched it. */
  customised: boolean;
  enabledByDefault: boolean;
  updatedBy: string | null;
}

/**
 * What this deployment has switched on, and what this user may therefore see.
 *
 * <p>Loaded once per session and held in a signal: the answer gates navigation and page sections, so
 * asking the server on every render would put a round trip in front of the menu.
 *
 * <p>Fails **open** for features that ship enabled and **closed** for the rest — see `enabled()`.
 * That is deliberate: if the flag list cannot be fetched, the application should still behave as it
 * did before flags existed rather than appearing to have lost half its features.
 */
@Injectable({ providedIn: 'root' })
export class FeatureService {
  private readonly base = `${environment.apiBase}/api/features`;
  private readonly auth = inject(AuthService);
  private readonly http = inject(HttpClient);

  /** Keys available to the signed-in user. Null until loaded. */
  private readonly available = signal<Set<FeatureKey> | null>(null);
  readonly loaded = computed(() => this.available() !== null);

  /**
   * Features that ship on. Used only as the fallback while the list is unknown, so a slow or failed
   * fetch degrades to "the application as it was" rather than to an empty shell.
   */
  private static readonly ON_BY_DEFAULT: ReadonlySet<string> = new Set([
    'VIDEO_LIBRARY',
    'VIDEO_ENGAGEMENT',
    'VIDEO_DOWNLOAD',
    'LOUNGE_CHAT',
    'AI_DRAFTING',
  ]);

  private headers(): Record<string, string> {
    const token = this.auth.token();
    return token ? { Authorization: `Bearer ${token}` } : {};
  }

  /** Ask what this user may use. Safe to call repeatedly; the result replaces the cache. */
  refresh(): Observable<string[]> {
    return this.http
      .post<string[]>(`${this.base}/my-features`, {}, {
        headers: this.headers(),
        // SILENT: bootstrap read on every page load; the app renders around it.
        context: new HttpContext().set(SILENT, true),
      })
      .pipe(tap((keys) => this.available.set(new Set(keys as FeatureKey[]))));
  }

  /** Forget the cache — call on sign-out, since the answer is per user. */
  clear(): void {
    this.available.set(null);
  }

  enabled(key: FeatureKey): boolean {
    const known = this.available();
    if (known === null) return FeatureService.ON_BY_DEFAULT.has(key);
    return known.has(key);
  }

  // ---- administration (ADMIN only; the server enforces it) -------------------

  list(): Observable<FeatureView[]> {
    return this.http.post<FeatureView[]>(`${this.base}/list-features`, {}, {
        headers: this.headers(),
        // SILENT: the feature admin table loading its own rows.
        context: new HttpContext().set(SILENT, true),
      });
  }

  assignableRoles(): Observable<string[]> {
    return this.http.post<string[]>(`${this.base}/assignable-roles`, {}, {
        headers: this.headers(),
        // SILENT: a dropdown's options, fetched beside the table.
        context: new HttpContext().set(SILENT, true),
      });
  }

  /**
   * Turn a feature on or off, and optionally change which roles may use it.
   *
   * Pass `null` for `allowedRoles` to leave the role configuration alone — the server reads null as
   * "not changing this". An empty array is different: it explicitly means "no role", and is stored.
   * Conflating the two made unticking the last role silently restore every default.
   */
  set(key: FeatureKey, enabled: boolean, allowedRoles: string[] | null): Observable<FeatureView> {
    return this.http
      .post<FeatureView>(
        `${this.base}/set-feature`,
        { key, enabled, allowedRoles },
        { headers: this.headers() },
      )
      // The change may affect the current user's own menu, so re-read what they can see.
      .pipe(tap(() => this.refresh().subscribe({ error: () => {} })));
  }

  /**
   * Apply one decision to every feature at once — the setup step for a fresh deployment.
   *
   * Granting every role to everything cannot escalate anything: roles narrow access and never widen
   * it, since the server checks Spring Security first and independently.
   */
  setAll(enabled: boolean, allowedRoles: string[] | null): Observable<FeatureView[]> {
    return this.http
      .post<FeatureView[]>(
        `${this.base}/set-all-features`,
        { enabled, allowedRoles },
        { headers: this.headers() },
      )
      .pipe(tap(() => this.refresh().subscribe({ error: () => {} })));
  }

  /** Clear every override, returning the deployment to how it ships. The way back from setAll. */
  resetAll(): Observable<FeatureView[]> {
    return this.http
      .post<FeatureView[]>(`${this.base}/reset-all-features`, {}, { headers: this.headers() })
      .pipe(tap(() => this.refresh().subscribe({ error: () => {} })));
  }

  reset(key: FeatureKey): Observable<FeatureView> {
    return this.http
      .post<FeatureView>(`${this.base}/reset-feature`, { key }, { headers: this.headers() })
      .pipe(tap(() => this.refresh().subscribe({ error: () => {} })));
  }
}
